Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29633 Support C++ tensor advanced indexing
* **#29632 Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified**

This PR is BC-breaking in the following way:

Previously, C++ `torch::tensor` with an integer type (e.g. `torch::tensor(1)`) or a (nested) braced-init-list of integer types (e.g. `torch::tensor({{1, 2}})`) produces a tensor with the same dtype. Now it always produces a tensor of dtype `torch::kLong` (aka. int64_t). matching Python `torch.tensor` behavior.

Differential Revision: [D18465819](https://our.internmc.facebook.com/intern/diff/D18465819)

Commit:
Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

kostmo commented:
## CircleCI build failures summary
As of commit 3016786:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=3016786c2ba36d9561408647e7ec606f3028888d).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

zou3519 commented on test/cpp/api/support.h:
```diff
@@ -61,5 +61,17 @@ inline int count_substr_occurrences(const std::string& str, const std::string& s
   return count;
 }
 
+// A RAII, thread local (!) guard that changes default dtype upon
```
Just checking, `torch::set_default_dtype` changes the default dtype locally?

zou3519 replied:
It seems global: https://github.com/pytorch/pytorch/blob/master/c10/core/DefaultDtype.cpp. Orthogonal to this PR, does this need synchronization?

zou3519 replied:
Oh sorry, I totally did not realize this was in test code. This is fine in testing code, I was wondering if it was OK as an API.

yf225 replied:
Thanks for the catch - I think this `AutoDefaultDtypeMode` guard in its original form was not safe to use if we ever run C++ tests in parallel in multiple threads, since a test in thread A might not have finished when another test in thread B changes the default dtype using `torch::set_default_dtype`. To address this problem, I made the following two changes:
1. Add `static std::mutex default_dtype_mutex` to `AutoDefaultDtypeMode`, so that usage of `AutoDefaultDtypeMode` is synchronized and thus always thread-safe.
2. Change other C++ tests that have `set_default_dtype(...)` to always use `AutoDefaultDtypeMode` to guard the whole scope of the test, so that access to `set_default_dtype(...)` is effectively synchronized *within the C++ test suite*. (It won't protect us against running a C++ test while changing default dtype through Python within the same process, but such usage seems more intentional than accidental, and the author of that test mechanism should be responsible for making sure such usage is sane.)

yf225 replied:
I think the `AutoDefaultDtypeMode` guard is pretty ad-hoc - for now it's only used for the C++ tests, and its RAII mechanism should work :D

zou3519 replied:
Does gtest run tests in parallel / is it able to?

yf225 replied:
I don't think gtest by itself can run tests in parallel (https://cuhkszlib-xiaoxing.readthedocs.io/en/latest/external/gtest/googletest/docs/FAQ.html#does-google-test-support-running-tests-in-parallel), but there are some test runners that can make it work: https://github.com/google/gtest-parallel

zou3519 replied:
Based on the design of the tests (they only run single-threaded), it might be better to simplify the data structure by removing the mutex and writing a quick comment that this isn't thread-safe and that it doesn't matter because we only use one thread

zou3519 commented:
Does it make sense to use the default dtype when the user does `torch.tensor(1.0)`?  1.0 is "always" a double when it's used in C++

zou3519 commented:
> Does it make sense to use the default dtype when the user does `torch.tensor(1.0)`? 1.0 is "always" a double when it's used in C++

On second thought, python floats also have double-precision and we use the default dtype for them.

zou3519 commented on test/cpp/api/tensor.cpp:
```diff
@@ -309,12 +318,31 @@ TEST(TensorTest, TorchTensorCtorSingleDimIntegralType) {
-TEST(TensorTest, TorchTensorCtorSingleDimFloatingType) {
+void test_TorchTensorCtorSingleDimFloatingType_expected_dtype(c10::ScalarType default_dtype) {
+  AutoDefaultDtypeMode dtype_mode(default_dtype);
+
```
For completeness I'd also like to see `auto tensor = torch::tensor({1.5f, 2.25f, 3.125f});`

yf225 replied:
Added

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -24,17 +24,22 @@ inline std::ostream& operator<<(std::ostream& stream, c10::BFloat16 value) {
+    // When `scalar_type == at::kDouble`, we know that the user is passing in
+    // a floating-point literal without specifying its type (e.g. `1.0` instead of `1.0f`).
+    // In Python, the dtype of `torch.tensor(1.0)` depends on the value of
+    // `torch.get_default_dtype()`, and we should do the same for C++ `torch::tensor(1.0)`.
```
It might be nice to have a parity test for this behavior. i.e., if we ever change the default tensor type behavior then we know to change it in both APIs. I don't see this changing anytime soon though so this is more of a nit.

yf225 replied:
Agreed - I will add it in a follow-up PR to improve the parity test mechanism.

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

Commit:
Update on "Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified"

Use default dtype for torch::tensor(floating_point_values) and torch::tensor(empty braced-init-list) when dtype is not specified

gh-metadata: pytorch pytorch 29632 gh/yf225/66/head

zou3519 reviewed (approved):
Looks good. I think it would be nice to remove the mutex to simplify the logic (since our tests are single-threaded anyways and probably won't change anytime soon) but let me know your thoughts

yf225 commented:
> Looks good. I think it would be nice to remove the mutex to simplify the logic (since our tests are single-threaded anyways and probably won't change anytime soon) but let me know your thoughts

Thanks! I feel that it's probably worth keeping the mutex because changing the global state back and forth in the tests does seems a bit worrying to me, and it might cause hard-to-debug flaky tests if we ever run them in parallel without the mutex.

zou3519 commented:
> > Looks good. I think it would be nice to remove the mutex to simplify the logic (since our tests are single-threaded anyways and probably won't change anytime soon) but let me know your thoughts
> 
> Thanks! I feel that it's probably worth keeping the mutex because changing the global state back and forth in the tests does seems a bit worrying to me, and it might cause hard-to-debug flaky tests if we ever run them in parallel without the mutex.

That's reasonable, let's keep it then.
